### DISCUSSION TITLE
INS: unable to initialise driver

### DISCUSSION THREAD

**Discussion by HefnySco** [HefnySco/OBAL#21] [Q&A] [answered]
Each time you start your OBAL you get this message:

**INS: unable to initialise driver**

**Comment by HefnySco** [comment, marked as answer]
This happens when one of four:

1. soldering are poor or you are testing with wires and jumpers that are not connected properly.
2. IMU is dead because of high temperature, bad polarity ...etc.
3. You are using the wrong software version.  Please note that OBALv1 & OBAL-GY91 each has its own version of code.
4. IMU is a clone and has a bad unknown ID. You can discover this easily using arduino & i2c connection and get REGID. Then you can edit [AP_InertialSensor_Invensense_registers.h](https://github.com/ArduPilot/ardupilot/blob/ba6cf1a42ba679819b1ab3dbab1622d777bc4e1b/libraries/AP_InertialSensor/AP_InertialSensor_Invensense_registers.h#L191) code by yourself the compi;le the code again.


Remember if you see [INS: unable to initialise driver] that means BARO is already detected.

**Reply by kar1150** [reply]
hello, could you please share a tip on how to fetch the id from arduino i2c?
I have already tried many arduino i2c libs, got the scan confirm 0x68 address and also got raw data from sensors, but unfortunately I can not get the ID.
Thank you @HefnySco amazing work you have done here!

**Reply by HefnySco** [reply]
You need to open the i2c lib that you use. you will find a function like this:

```
uint8_t MPU6050::getDeviceID() {
    I2Cdev::readBits(devAddr, MPU6050_RA_WHO_AM_I, MPU6050_WHO_AM_I_BIT, MPU6050_WHO_AM_I_LENGTH, buffer);
    return buffer[0];
}
```

the value of **buffer[0];** is what you need to know and update the ardupilot code accordingly.
in my case I had to add:
`#define MPU_WHOAMI_MPU9255_1    0x75
`

**Comment by highfreq** [comment]
Should i be able to see GY-91 when running i2cdetect -y 1 ?

Now i only see 40 and 48 which i guess are the boards PCA9685 and ADS1115

**Reply by DroneEngage** [reply]
GY-91 is connected via SPI you cannot see it using i2cdetect unless you are connecting it via i2c

**Reply by highfreq** [reply]
I understand, i can see the baro in mission planner but i still get the initialization error. I downloaded ardupilot than i patched without error and compiled, so it shouldn't be a software problem but i still can't see gyro and have baro error.

**Comment by highfreq** [comment]
The only thing i can see is inside mission planner on HW ID is a baro DevID 196866 Bus SPI address 1 Baro_BMP280.

But no gyro accel and compass (9250).

**Comment by highfreq** [comment]
During compilation i used ./waf --board obal   should i have used something different to compile for the gy-91?

**Comment by highfreq** [comment]
Please let me know if what i did is correct:

git clone https://github.com/ArduPilot/ardupilot.git
cd ardupilot 
git submodule update --init  --recursive
wget https://raw.githubusercontent.com/HefnySco/OBAL/main/Software/patches/obal_gy91.patch
git apply obal_gy91.patch
./waf configure --board obal
./waf rover

**Comment by highfreq** [comment]
Do i have to set something when enabling SPI???

I tested 4 different GY-91 and all of them have the initialize problem.

**Reply by DroneEngage** [reply]
I do not think so. I normally power it from the USB. but try to power it from a power supply and login ssh remotely using wifi and see what happens.

**Comment by highfreq** [comment]
Could it be a problem from the fact i power the board from the raspberry usb?
